activesupport/lib/active_support/cache/streaming_compressor.rb

```diff
@@ -0,0 +1,110 @@
+# frozen_string_literal: true
+
+require "zlib"
+require "concurrent/executor/cached_thread_pool"
+require "active_support/core_ext/numeric/bytes"
+
+module ActiveSupport
+  module Cache
+    # Serializes object and compresses serialized representation in streaming way.
+    # This avoids storing whole copy of serialized intermediate representation,
+    # like this happens in `Zlib.deflate(Marshal.dump(value))`.
+    module StreamingCompressor
+      DEFLATE_BUFFER_SIZE = 16.kilobytes.to_i
+      INFLATE_CHUNK_SIZE = 4.kilobyte.to_i
+
+      extend self
+
+      # Dumps value into string. Returns nil if `:compress_threshold` option
+      # is set and serialized value is less than given value.
+      def dump(value, **options)
+        io = StringIO.new
+        raw_bytes = dump_to_io(value, io, **options)
+        return unless raw_bytes
+        io.rewind
+        result = io.read
+        result if result.bytesize < raw_bytes
+      end
+
+      # Same as dump but streams compressed data. Returns nil and don't stream
+      # anything if `:compress_threshold` option is set
+      # and serialized value is less than given value.
+      def dump_to_io(value, to_io, compress_threshold: nil)
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            Marshal.dump(value, wio)
+          ensure
+            wio.close
+          end
+        end
+        if compress_threshold
+          chunk = rio.read([DEFLATE_BUFFER_SIZE, compress_threshold].max)
```

Comment:
Let's get rid of this extra array allocation with the `max` calculation

```ruby
size = DEFLATE_BUFFER_SIZE
size = compress_threshold if compress_threshold > size
chunk = rio.read([DEFLATE_BUFFER_SIZE, compress_threshold].max)
```

Reply:
`compress_threshold && compress_threshold > DEFLATE_BUFFER_SIZE ? size = compress_threshold : size = DEFLATE_BUFFER_SIZE`
Removes the `if` and separate assigning to `size` in multiple lines.
Cleaner, personal opinion though.

Reply:
I'm affraid this may not work with current tests which has compress_threshold: 1. Inflate definitely does not work with  such small chunks. Will check deflate.

Reply:
Ruby 2.4 (min version for Rails 6) includes [an optimization](https://bugs.ruby-lang.org/issues/12172) which avoids allocating an extra array for certain invocations of `[].max`, which I think includes this case.

Reply:
@schneems rails master requires ruby >=2.4.1, which compiles this call to `opt_newarray_max 2`, instead of `newarray 2; 0015 opt_send_without_block <callinfo!mid:max...` in previous versions. Should I update the code?